src/commands/utils/host.ts

```diff
@@ -20,3 +22,12 @@ export function getExtensionPath(): string | undefined {
     }
     return vscodeExtensionPath;
 }
+
+export function registerCommandWithTelemetry(
+    commandId: string,
+    callback: any
+) {
+
+    reporter.sendTelemetryEvent("command", { command: commandId });
```

Comment:
No, this is wrong.  This will send an event for every command when the user activates the extension, and then do nothing when the user _invokes_ the command.

What you need is something more like:

```
function registerCommand(command: string, callback: (...args: any[]) => any): vscode.Disposable {
    const wrappedCallback = telemetrise(command, callback);
    return vscode.commands.registerCommand(command, wrappedCallback);
}

function telemetrise(command: string, callback: (...args: any[]) => any): (...args: any[]) => any {
    return (a) => {
        if (reporter) {
            reporter.sendTelemetryEvent("command", { command: command });
        }
        return callback(a);
    };
}
```

This is a simplified version of the `vscode-kubernetes-tools` wrapping code - you may need to make some changes to make it work with the `vscode-azureextensionui` stuff.  But the concept remains the same:

* The `telemetrise` function takes a command handler, and produces a new handler that does _exactly the same thing_ except it shoots off some telemetry first.
* The `registerCommand` function does NOT send telemetry events itself.  Rather, it asks `telemetrise` to give it an instrumented version of the handler.  Then it registers that instrumented handler instead of the vanilla handler.

Consequently, when the user issues a command registered in this way, VS Code hands control to the _instrumented_ handler (because that's the one we registered), and that handler queues a telemetry event, and then hands off to the "original" handler.

Note also that we do some typing to ensure that callers of `registerCommad` only pass legitimate handlers, and that we have to take care to forward args from the instrumented handler to the underlying original handler.

Reply:
Ah, HOFs :)